DataTable aggregate rows feature

eadams17 opened the pull request:
## WHAT
add functionality to the `DataTable` component to handle aggregate row data

## WHY
we need this in preparation for the admin Growth Diagnostic reports

## HOW
update logic in the `DataTable` component to check for and handle aggregate row data and add an example to Backpack

just a note: I went with the name `aggregate_rows` for the identifier. @anathomical, I know we had said `break_outs`, but I felt like this made more sense. I'm totally open to changing it back, or to any other naming convention that @emilia-friedberg might feel like works better. I also plan on adding more thorough testing once the reports are actually implemented since this functionality may end up changing

### Screenshots
(If applicable. Also, please censor any sensitive data)
<img width="1054" alt="Screen Shot 2023-09-25 at 4 49 00 PM" src="https://github.com/empirical-org/Empirical-Core/assets/25959584/b484d185-7a05-45b9-b4e3-6a0c5a0bc2ff">
<img width="1056" alt="Screen Shot 2023-09-25 at 5 11 49 PM" src="https://github.com/empirical-org/Empirical-Core/assets/25959584/dba8ba7a-6510-40c4-855c-083dd0bfeb14">

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)
https://www.notion.so/quill/New-Project-Card-Admin-Diagnostic-Reports-Spec-For-Three-Reports-15ace9cefb7c4bf7b5475b041fff56d9?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? | yes
Have you deployed to Staging? | yes
Self-Review: Have you done an initial self-review of the code below on Github? | yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | yes

eadams17 committed:
add logic to DataTable component to handle aggregate row data

eadams17 committed:
reconcile develop

eadams17 committed:
add icons; remove test data

eadams17 committed:
add tests

eadams17 committed:
add example to Backpack

eadams17 committed:
fix linter errors, tweak variable naming; update snapshots

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Shared/components/shared/dataTable.tsx:
```diff
@@ -340,10 +367,24 @@ export class DataTable extends React.Component<DataTableProps, DataTableState> {
+    if (shouldDisplayAggregateIcon) {
+      aggregateRowButtonOrIcon = <img alt="" className="aggregate-row-icon" src={aggregateRowArrowSrc} />
+    } else if (shouldDisplayToggleButton && aggregateRowsDisplayed) {
+      aggregateRowButtonOrIcon = <button className="aggregate-row-toggle interactive-wrapper focus-on-light" onClick={this.handleHideAggregateRows}><img alt="" className="aggregate-row-toggle-icon" src={toggleArrowExpandedSrc} /></button>
```
This button, and the next one, need an aria-label so that screenreader users know what this button does.

eadams17 replied:
will do! thanks for calling out

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Shared/components/shared/dataTable.tsx:
```diff
@@ -376,17 +418,31 @@ export class DataTable extends React.Component<DataTableProps, DataTableState> {
-
+    let rowElement = <tr className={rowClassName} key={String(row.id)}>{rowContent}</tr>
+    const aggregateRowIdentifier = `${row[headers[0].attribute]}-${row.id}`
+    const showAggregateRows = row.aggregate_rows && row.aggregate_rows.length && expandedParentRowIdentifier && expandedParentRowIdentifier === aggregateRowIdentifier
```
I think this could just be shortened to `const showAggregateRows = row.aggregate_rows?.length && expandedParentRowIdentifier === aggregateRowIdentifier`, because neither of those calls should error so the initial null checks aren't necessary.

emilia-friedberg commented on services/QuillLMS/client/app/bundles/Shared/components/shared/dataTable.tsx:
```diff
@@ -340,10 +367,24 @@ export class DataTable extends React.Component<DataTableProps, DataTableState> {
+    const shouldDisplayAggregateIcon = isAggregateRow && i === 0
+    const shouldDisplayToggleButton = row.aggregate_rows && row.aggregate_rows.length && i === 0
+    const aggregateRowsDisplayed = expandedParentRowIdentifier && expandedParentRowIdentifier === aggregateRowIdentifier
+    let aggregateRowButtonOrIcon = <span />
```
could this line just be `let aggregateRowButtonOrIcon` without setting it equal to `<span />`? It doesn't really matter visually but it does create pretty busy HTML (as evidence by all those snapshot updates). I know sometimes the null element thing creates weird behavior though so if it doesn't work to remove it it's all good.

eadams17 replied:
yeah, that's a good point-- I'll make this change, and if there is any weird behavior down the line, I'll just revert to using a span

emilia-friedberg reviewed (approved):
I like this implementation a lot. I have some minor tweaks requested around concision and accessibility, but nothing I need to review again. Thanks for taking care of this!

emilia-friedberg commented:
@eadams17 one more question -- I think we'd been planning on making some of the generic styling changes to `DataTable` as part of this project (with the new header background color, etc). Is that still the case?

anathomical reviewed (approved):
This looks reasonable to me.  I wonder if it's worth adding some additional tests to the `dataTable.test.tsx` file.  We've got React testing in use, so maybe checking to make sure that things toggle as expected via tests is worth doing if that's straightforward is worth spending a bit of time on?

eadams17 commented:
> @eadams17 one more question -- I think we'd been planning on making some of the generic styling changes to `DataTable` as part of this project (with the new header background color, etc). Is that still the case?

that will be out of scope for this project, but I think the plan is still to eventually do a universal overhaul with the changes (I'll check in with Jack about it)

eadams17 commented:
> This looks reasonable to me. I wonder if it's worth adding some additional tests to the `dataTable.test.tsx` file. We've got React testing in use, so maybe checking to make sure that things toggle as expected via tests is worth doing if that's straightforward is worth spending a bit of time on?

for sure-- I made a note though in the How section about my plan to add more tests once we actually implement this toggle functionality with the new reports as we will probably end up having to make some tweaks to it

eadams17 committed:
small revisions

eadams17 committed:
update conditional statement

eadams17 committed:
Merge branch 'develop' of https://github.com/empirical-org/Empirical-Core into data-table-aggregate-rows-feature

eadams17 committed:
ADGR/ Overview frontend (#11114)

* add header for diagnostic reports overview

* add icons to buttons

* add tabs, dropdown and table

* update fill in the blanks concept results file to use shared function (#11109)

* replace diagnostic-grey icon (#11110)

* add more dropdown options

* tweak CSS

* update active states for buttons

* add tests

* fix linter errors

* update snapshot

* Change styling and copy on Diagnostic Reports pages (#11104)

* change to recommended practice

* try appending description

* try another way of stacking paragraphs

* try another way

* add one more break for visual matching

* snapshot update

* Style changes for shifted bigger element on progress page

* Reverse tooltip changes

* particularize style changes to only affect the cards in post diagnstic

* add card class

* Re-style Empty Progress Reports page (#11107)

* html and css changed

* make text bolder on link

* Update CMS ruby version to Ruby 3.2.0 (#11108)

* update gemfile and ruby versions

* update circleCI version

* Add rake task for updating missing labels (#11113)

* Add rake task for updating missing labels

* Activate model after update

* Add some logging

* Lms/admin snapshot front end retry (#11103)

* Testing in staging with 1 second retry

* WIP add some testing console logs to understand how this is working

* More testing

* WIP more testing

* Hopefully-final version of the code

* Add some logging

* Add more logging

* More logging

* Even more logging

* Pass retryTimeout through to getData

* More logging, trying to figure out encapsulation

* Allow for set calls on useState values to be async

* Hopefully final update with working clearTimeout

* Still trying to figure out clearTimeout

* Still chasing down timing issue

* Remove final logging

* Remember to update Rankings to retry with the same logic

* Lms/send pusher in sidekiq job (#11105)

* Create a Sidekiq worker to send Pusher messages

So that we can de-couple Pusher errors and retries from the rest of jobs that utilize Pusher

* Refactor Admin Snapshot workers to use Pusher Sidekiq worker

Instead of sending messages directly

* Refactor classroom import workers to use Pusher Sidekiq job

Instead of sending Pusher messages directly from the import workers.  This lets us decouple errors and retries triggered by Pusher from the actual work of the importers

* Fix linting error

* add CSS comments

* fix linter

---------

Co-authored-by: Emilia Friedberg <emiliafriedberg@gmail.com>
Co-authored-by: cissyxyu <57366100+cissyxyu@users.noreply.github.com>
Co-authored-by: brendanshean <brendanshean@gmail.com>
Co-authored-by: Anathomical <thomas.robertson@quill.org>
